### PR TITLE
[Android] Tasks and TaskLists are deleteing actually.

### LINKED ISSUE #299: Tasks and TaskLists on Android are not deleting when swiped
When swiping to delete a `TaskList` the object is just removed from the `TaskListList` array but not actually deleted. Likewise when swiping to delete a `Task` it is just removed from the `TaskList` array and not deleted.

This only appears on Android.

### PR DESCRIPTION
**Pull request by dalinaum**
Tasks are TasksLists are not deleting actually. They are removed from its array instead. This commit will fix this problem. Please review this @realm/java @bigfish24 

Fix #299

### CODE REVIEW

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskAdapter.java [hunk 1]
```diff
@@ -107,7 +107,9 @@ public void onItemDismissed(final int position) {
         realm.executeTransaction(new Realm.Transaction() {
             @Override
             public void execute(Realm realm) {
+                final Task task = getData().get(position);
                 getData().remove(position);
```
Comment (zaki50) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskAdapter.java]: Do we need this? I think `task.deleteFromRealm();` removes it from the RealmList

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskAdapter.java [hunk 2]
```diff
@@ -122,7 +124,9 @@ public void onItemReverted() {
         realm.executeTransaction(new Realm.Transaction() {
             @Override
             public void execute(Realm realm) {
+                final Task task = getData().get(0);
                 getData().remove(0);
```
Comment (zaki50) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskAdapter.java]: Same

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskListAdapter.java [hunk 3]
```diff
@@ -103,7 +102,9 @@ public void onItemDismissed(final int position) {
         realm.executeTransaction(new Realm.Transaction() {
             @Override
             public void execute(Realm realm) {
+                final TaskList taskList = getData().get(position);
                 getData().remove(position);
```
Comment (zaki50) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskListAdapter.java]: Same

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskListAdapter.java [hunk 4]
```diff
@@ -118,7 +119,9 @@ public void onItemReverted() {
         realm.executeTransaction(new Realm.Transaction() {
             @Override
             public void execute(Realm realm) {
+                final TaskList taskList = getData().get(0);
                 getData().remove(0);
```
Comment (zaki50) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskListAdapter.java]: Same

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskListAdapter.java [hunk 5]
```diff
@@ -103,7 +102,9 @@ public void onItemDismissed(final int position) {
         realm.executeTransaction(new Realm.Transaction() {
             @Override
             public void execute(Realm realm) {
+                final TaskList taskList = getData().get(position);
                 getData().remove(position);
+                taskList.deleteFromRealm();
```
Comment (zaki50) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskListAdapter.java]: When deleting a `TaskList`, I think we need to delete containing tasks as well.

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskListAdapter.java [hunk 6]
```diff
@@ -118,7 +119,9 @@ public void onItemReverted() {
         realm.executeTransaction(new Realm.Transaction() {
             @Override
             public void execute(Realm realm) {
+                final TaskList taskList = getData().get(0);
                 getData().remove(0);
+                taskList.deleteFromRealm();
```
Comment (zaki50) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskListAdapter.java]: When deleting a `TaskList`, I think we need to delete containing tasks as well.

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskAdapter.java (hunk 1 above)
Comment (dalinaum) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskAdapter.java]: I see. I am deleting `RealmList.delete(int)`.

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskListAdapter.java (hunk 5 above)
Comment (dalinaum) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/list/TaskListAdapter.java]: Right. I just checked that we should delete containing tasks as well.

**Comment by dalinaum** [comment]
Updated. Please review this @realm/java @zaki50

**Comment by zaki50** [comment]
I tried this branch.
When I delete `TaskList` object, other client which is opening the same remote realm file crashed and got following log :

```
W/InputDispatcher(  312): channel '53762098 io.realm.realmtasks/io.realm.realmtasks.TaskListActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
E/InputDispatcher(  312): channel '53762098 io.realm.realmtasks/io.realm.realmtasks.TaskListActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
I/ActivityManager(  312): Process io.realm.realmtasks (pid 1262) has died.
W/ActivityManager(  312): Force removing ActivityRecord{5376cf08 io.realm.realmtasks/.TaskListActivity}: app died, no saved state
W/InputDispatcher(  312): Attempted to unregister already unregistered input channel '53762098 io.realm.realmtasks/io.realm.realmtasks.TaskListActivity (server)'
I/WindowState(  312): WIN DEATH: Window{53762098 io.realm.realmtasks/io.realm.realmtasks.TaskListActivity paused=false}
I/WindowManager(  312): WINDOW DIED Window{53762098 io.realm.realmtasks/io.realm.realmtasks.TaskListActivity paused=false}
I/qtaguid (  312): Failed write_ctrl(s 0 10048) res=-1 errno=1
W/NetworkManagementSocketTagger(  312): setKernelCountSet(10048, 0) failed with errno -1
I/qtaguid (  312): Failed write_ctrl(s 1 10021) res=-1 errno=1
W/NetworkManagementSocketTagger(  312): setKernelCountSet(10021, 1) failed with errno -1
W/EGL_genymotion(  499): eglSurfaceAttrib not implemented
```

Realm Browser also crashed.

**Comment by cmelchior** [comment]
There should be more info than that? Either on the server log or in LogCat.
Although logging might not work fully in Java yet

**Comment by zaki50** [comment]
server log is 

```
2016-09-30T11:37:42.877Z - info: Logging to console at level 'info'.
2016-09-30T11:37:43.022Z - info: Realm Object Server web server listening on 127.0.0.1:27080.
2016-09-30T11:37:43.024Z - info: http proxy listening on 0.0.0.0:9080.
2016-09-30T11:37:43.128Z - info: The Realm Object Server requires an initial admin user to start. To continue, please open your web browser to access the dashboard at: http://127.0.0.1:27080 and register your user.
2016-09-30T11:38:01.287Z - info: Realm Object Server sync server listening on 127.0.0.1:27800.
2016-09-30T11:38:01.289Z - info: sync-server: Realm sync server started ([realm-core-2.0.0], [realm-sync-1.0.0-BETA-1.0])
2016-09-30T11:38:01.289Z - info: sync-server: Directory holding persistent state: /Users/zaki/Downloads/realm-mobile-platform/realm-object-server/object-server/root_dir/user_data
2016-09-30T11:38:01.289Z - info: sync-server: Listening on 127.0.0.1:27800 (sync protocol version 14)
2016-09-30T11:38:11.802Z - info: sync-server: Connection[1]: Connection from 127.0.0.1:64010
2016-09-30T11:38:11.802Z - info: sync-server: Connection[1]: Received: CLIENT(protocol_version=14, client_info_size=2, client_info = {})
2016-09-30T11:38:11.802Z - info: sync-server: Connection[1]: Session[1]: Initiated
2016-09-30T11:38:11.803Z - info: sync-server: Connection[1]: Session[1]: Received: BIND(server_path='/__admin', signed_user_token=eyJpZGVudGl0eSI6ImF1..., need_file_ident_pair=1)
2016-09-30T11:38:11.869Z - info: sync-server: Connection[1]: Session[1]: Sending: ALLOC(server_file_ident=8516832057658563069, client_file_ident=1, client_file_ident_secret=3568542495883050809)
2016-09-30T11:38:11.870Z - info: sync-server: Connection[1]: Session[1]: Received: IDENT(server_file_ident=8516832057658563069, client_file_ident=1, client_file_ident_secret=3568542495883050809, scan_server_version=0, scan_client_version=0, latest_server_version=0, latest_server_session_ident=0)
2016-09-30T11:38:24.321Z - info: sync-server: Connection[2]: Connection from 127.0.0.1:64277
2016-09-30T11:38:24.324Z - info: sync-server: Connection[2]: Received: CLIENT(protocol_version=14, client_info_size=2, client_info = {})
2016-09-30T11:38:24.324Z - info: sync-server: Connection[2]: Session[1]: Initiated
2016-09-30T11:38:24.325Z - info: sync-server: Connection[2]: Session[1]: Received: BIND(server_path='/cf243a650bf16ad6acc53d43363f7c08/realmtasks', signed_user_token=eyJpZGVudGl0eSI6ImNm..., need_file_ident_pair=1)
2016-09-30T11:38:24.351Z - info: sync-server: Connection[2]: Session[1]: Sending: ALLOC(server_file_ident=6243504645818668562, client_file_ident=1, client_file_ident_secret=7055788955003076757)
2016-09-30T11:38:24.362Z - info: sync-server: Connection[2]: Session[1]: Received: IDENT(server_file_ident=6243504645818668562, client_file_ident=1, client_file_ident_secret=7055788955003076757, scan_server_version=0, scan_client_version=0, latest_server_version=0, latest_server_session_ident=0)
2016-09-30T11:38:29.128Z - info: sync-server: Connection[3]: Connection from 127.0.0.1:64391
2016-09-30T11:38:29.129Z - info: sync-server: Connection[3]: Received: CLIENT(protocol_version=14, client_info_size=2, client_info = {})
2016-09-30T11:38:29.129Z - info: sync-server: Connection[3]: Session[1]: Initiated
2016-09-30T11:38:29.130Z - info: sync-server: Connection[3]: Session[1]: Received: BIND(server_path='/cf243a650bf16ad6acc53d43363f7c08/realmtasks', signed_user_token=eyJpZGVudGl0eSI6ImNm..., need_file_ident_pair=1)
2016-09-30T11:38:29.131Z - info: sync-server: Connection[3]: Session[1]: Sending: ALLOC(server_file_ident=6243504645818668562, client_file_ident=2, client_file_ident_secret=4834737374322599508)
2016-09-30T11:38:29.134Z - info: sync-server: Connection[3]: Session[1]: Received: IDENT(server_file_ident=6243504645818668562, client_file_ident=2, client_file_ident_secret=4834737374322599508, scan_server_version=0, scan_client_version=0, latest_server_version=0, latest_server_session_ident=0)
2016-09-30T11:38:52.251Z - info: sync-server: Connection[3]: Session[1]: Terminated
2016-09-30T11:38:52.251Z - info: sync-server: Connection[3]: Connection closed by client: End of input
```

RealmBrowser is not connected at this time.

**Comment by cmelchior** [comment]
Currently the warning level is hardcoded ( @beeender is working on that), but if you change https://github.com/realm/realm-java/blob/master/realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp#L103 to `trace` instead you should get full output from Sync

**Comment by zaki50** [comment]
@dalinaum Can you reproduce this? I'm using 2 genymotion devices.

**Comment by nhachicha** [comment]
I can confirm zaki's bug 
```
E/AndroidRuntime( 1558): io.realm.exceptions.RealmError: Unrecoverable error. Bad transaction log in /Users/emanuelez/repos/realm/realm-java-private/realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp line 123
E/AndroidRuntime( 1558):       	at io.realm.SyncManager.nativeRunClient(Native Method)
E/AndroidRuntime( 1558):       	at io.realm.SyncManager.access$000(SyncManager.java:46)
E/AndroidRuntime( 1558):       	at io.realm.SyncManager$2.run(SyncManager.java:100)
E/AndroidRuntime( 1558):       	at java.lang.Thread.run(Thread.java:841)
W/ActivityManager(  548):   Force finishing activity io.realm.realmtasks/.TaskListActivity
```

**Comment by zaki50** [comment]
```
D/REALM   ( 1951): AuthenticateResponse. Identity ad05aab64f042270e4d1c02fc99f239c; Path /ad05aab64f042270e4d1c02fc99f239c/realmtasks
D/REALM   ( 1951): Session[/data/data/io.realm.realmtasks/files/realm-object-server/ad05aab64f042270e4d1c02fc99f239c/realmtasks]: Access token acquired
D/REALM   ( 1951): Session[realm://10.0.1.51:9080/ad05aab64f042270e4d1c02fc99f239c/realmtasks]: AUTHENTICATING -> BINDING
V/REALM   ( 1951):  --> Java_io_realm_internal_objectserver_SyncSession_nativeBind
D/REALM   ( 1951): Session[realm://10.0.1.51:9080/ad05aab64f042270e4d1c02fc99f239c/realmtasks]: BINDING -> BOUND
I/REALM   ( 1951): Opening Realm file: /data/data/io.realm.realmtasks/files/realm-object-server/ad05aab64f042270e4d1c02fc99f239c/realmtasks
D/REALM   ( 1951): Connection[1]: WebSocket::Websocket()
D/REALM   ( 1951): Connection[1]: Reconnecting in 0 milliseconds
I/REALM   ( 1951): Connection[1]: Session[1]: Starting session for '/data/data/io.realm.realmtasks/files/realm-object-server/ad05aab64f042270e4d1c02fc99f239c/realmtasks'
D/REALM   ( 1951): Connection[1]: Session[1]: last_version_available  = 17
D/REALM   ( 1951): Connection[1]: Session[1]: progress_server_version = 12
D/REALM   ( 1951): Connection[1]: Session[1]: progress_client_version = 8
I/REALM   ( 1951): Connection[1]: Resolving '10.0.1.51:9080'
I/REALM   ( 1951): Connection[1]: Connecting to endpoint '10.0.1.51:9080' (1/1)
I/REALM   ( 1951): Connection[1]: Connected to endpoint '10.0.1.51:9080' (from '10.0.3.15:35733')
D/REALM   ( 1951): Connection[1]: WebSocket::initiate_client_handshake()
V/REALM   ( 1951): Connection[1]: HTTP request =
V/REALM   ( 1951): GET /realm-object-server HTTP/1.1
V/REALM   ( 1951): Host: 10.0.1.51
V/REALM   ( 1951): Upgrade: websocket
V/REALM   ( 1951): Connection: Upgrade
V/REALM   ( 1951): Sec-WebSocket-Key: AehAHidraSjKIZZHn6Su9A==
V/REALM   ( 1951): Sec-WebSocket-Version: 13
V/REALM   ( 1951): Sec-WebSocket-Protocol: io.realm.protocol
V/REALM   ( 1951):
V/REALM   ( 1951): Connection[1]: WebSocket::read_http_message_line()
V/REALM   ( 1951): Connection[1]: Websocket: http_line = HTTP/1.1 101 Switching Protocols
V/REALM   ( 1951): Connection[1]: WebSocket::read_http_message_line()
V/REALM   ( 1951): Connection[1]: Websocket: http_line = upgrade: websocket
V/REALM   ( 1951): Connection[1]: WebSocket::read_http_message_line()
V/REALM   ( 1951): Connection[1]: Websocket: http_line = connection: Upgrade
V/REALM   ( 1951): Connection[1]: WebSocket::read_http_message_line()
V/REALM   ( 1951): Connection[1]: Websocket: http_line = sec-websocket-accept: 5dxEuIB2yRJuTRsc4JiapE3Dr2Y=
V/REALM   ( 1951): Connection[1]: WebSocket::read_http_message_line()
V/REALM   ( 1951): Connection[1]: Websocket: http_line = sec-websocket-protocol: io.realm.protocol
V/REALM   ( 1951): Connection[1]: WebSocket::read_http_message_line()
V/REALM   ( 1951): Connection[1]: Websocket: http_line =
V/REALM   ( 1951): Connection[1]: WebSocket::handle_http_response_received()
I/REALM   ( 1951): Connection[1]: Sending: CLIENT(protocol_version=14, client_info_size=2)
V/REALM   ( 1951): Connection[1]: WebSocket::frame_reader_loop()
V/REALM   ( 1951): Connection[1]: stage_init
V/REALM   ( 1951): Connection[1]: WebSocket: frame_reader.read_size = 2
I/REALM   ( 1951): Connection[1]: Session[1]: Sending: BIND(server_path='/ad05aab64f042270e4d1c02fc99f239c/realmtasks', signed_user_token_size=605, need_file_ident_pair=0)
V/REALM   ( 1951): Connection[1]: handle_write_output_buffer
I/REALM   ( 1951): Connection[1]: Session[1]: Sending: IDENT(server_file_ident=3741612703900327660, client_file_ident=2, client_file_ident_secret=7707227788121299416, scan_server_version=12, scan_client_version=8, latest_server_version=12, latest_server_session_ident=4053167523229729553)
V/REALM   ( 1951): Connection[1]: handle_write_output_buffer
D/REALM   ( 1951): Using already open Realm file: /data/data/io.realm.realmtasks/files/realm-object-server/ad05aab64f042270e4d1c02fc99f239c/realmtasks
V/REALM   ( 1951): Connection[1]: WebSocket::frame_reader_loop()
V/REALM   ( 1951): Connection[1]: stage_header_beginning
V/REALM   ( 1951): Connection[1]: WebSocket: frame_reader.read_size = 106
V/REALM   ( 1951): Connection[1]: WebSocket::frame_reader_loop()
D/REALM   ( 1951): Connection[1]: stage_payload
D/REALM   ( 1951): Connection[1]: websocket message received: download 1 13 8 13 4053167523229729553 8 0 58 0
D/REALM   ( 1951): 13 8 55169379340 1 36
D/REALM   ( 1951): Connection[1]: message_type = download
D/REALM   ( 1951): Connection[1]: Session[1]: Received: DOWNLOAD(scan_server_version=13, scan_client_version=8, latest_server_version=13, latest_server_session_ident=4053167523229729553, latest_client_version=8, is_body_compressed=0, uncompressed_body_size=58, compressed_body_size=0)
V/REALM   ( 1951): Connection[1]: Session[1]: Received: DOWNLOAD CHANGESET(server_version=13, client_version=8, origin_timestamp=55169379340, origin_client_file_ident=1, changeset_size=36)
V/REALM   ( 1951): Connection[1]: Session[1]: Changeset: 05 00 02 0E 01 01 02 01 05 00 03 1E 03 01 02 24 00 01 05 00 04 1E 01 00 03 24 00 01 05 00 03 0E 01 01 02 01
D/REALM   ( 1951): Using already open Realm file: /data/data/io.realm.realmtasks/files/realm-object-server/ad05aab64f042270e4d1c02fc99f239c/realmtasks
E/REALM   ( 1951): jni: ThrowingException 5, Bad transaction log in /Users/zaki/fromgit/realm/realm-java/realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp line 123, .
E/REALM   ( 1951): jni: ThrowingException 5, Bad transaction log in /Users/zaki/fromgit/realm/realm-java/realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp line 123, .
W/dalvikvm( 1951): threadid=10: thread exiting with uncaught exception (group=0xa6222288)
E/AndroidRuntime( 1951): FATAL EXCEPTION: RealmSyncClient
E/AndroidRuntime( 1951): io.realm.exceptions.RealmError: Unrecoverable error. Bad transaction log in /Users/zaki/fromgit/realm/realm-java/realm/realm-library/src/main/cpp/io_realm_SyncManager.cpp line 123
E/AndroidRuntime( 1951): 	at io.realm.SyncManager.nativeRunClient(Native Method)
E/AndroidRuntime( 1951): 	at io.realm.SyncManager.access$000(SyncManager.java:46)
E/AndroidRuntime( 1951): 	at io.realm.SyncManager$2.run(SyncManager.java:100)
E/AndroidRuntime( 1951): 	at java.lang.Thread.run(Thread.java:856)
```

**Comment by zaki50** [comment]
I confirmed that Realm-Java 2.0.2 solved above problem. No crash happens anymore.

**Comment by dalinaum** [comment]
Please review this @realm/java

**Pull request merged**